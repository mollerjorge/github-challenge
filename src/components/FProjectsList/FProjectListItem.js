import React from "react";
import PropTypes from "prop-types";
import FProjectsListItemHolder from "./FProjectListItem.style.js";
import { Icon } from "antd";

const FProjectsListItem = ({
	project: { name, id, watchers_count },
	onClickProjectHandler
}) => {
	return (
		<FProjectsListItemHolder
			onClick={onClickProjectHandler.bind(this, id, name)}
		>
			<h4>{name}</h4>{" "}
			<div>
				<Icon type="eye" /> <span>{watchers_count}</span>
			</div>
		</FProjectsListItemHolder>
	);
};

FProjectsListItem.propTypes = {
	project: PropTypes.shape({
		name: PropTypes.string,
		id: PropTypes.number,
		watchers_count: PropTypes.number
	}),
	onClickProjectHandler: PropTypes.func.isRequired
};

export default FProjectsListItem;
