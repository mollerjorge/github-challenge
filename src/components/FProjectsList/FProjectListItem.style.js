import styled from "styled-components";

const FProjectListItemHolder = styled.div`
	&& {
		background: #ffffff;
		cursor: pointer;
		border-radius: 0.1rem;
		display: flex;
		padding: 1rem;
		justify-content: space-between;
		margin: 0.5rem 1rem;
		transition: all 0.3s ease-in-out;

		&:hover {
			transform: scale(1.05);
		}
		h4 {
			word-wrap: break-word;
			max-width: 80%;
		}
	}
`;

FProjectListItemHolder.displayName = "FProjectsListItemHolder";

export default FProjectListItemHolder;
