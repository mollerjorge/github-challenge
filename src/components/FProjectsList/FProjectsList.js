import React from "react";
import PropTypes from "prop-types";
import FProjectListItem from "./FProjectListItem";

const FProjectsList = ({ projects, onClickProjectHandler, searchValue }) => {
  const displayProjectsList = projects
    .sort((p1, p2) => p2.watchers_count - p1.watchers_count)
    .map(p => {
      if (p.name.toLowerCase().includes(searchValue.toLowerCase())) {
        return (
          <FProjectListItem
            onClickProjectHandler={onClickProjectHandler}
            key={p.id}
            project={p}
          />
        );
      } else {
        return <React.Fragment key={p.id} />;
      }
    });
  return <React.Fragment>{displayProjectsList}</React.Fragment>;
};

FProjectsList.propTypes = {
  projects: PropTypes.array.isRequired,
  onClickProjectHandler: PropTypes.func.isRequired,
  searchValue: PropTypes.string
};

export default FProjectsList;
