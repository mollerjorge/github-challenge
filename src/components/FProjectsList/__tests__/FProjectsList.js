import React from "react";
import { shallow } from "enzyme";
import FProjectsList from "../FProjectsList";

describe("FProjectsList", () => {
	const mockProjectsList = [
		{
			name: "create-react-app",
			id: 1,
			watchers_count: 1000
		},
		{
			name: "flow",
			id: 2,
			watchers_count: 500
		}
	];
	const mockFn = jest.fn();
	const fProjectList = shallow(<FProjectsList projects={mockProjectsList} onClickProjectHandler={mockFn} searchValue=""/>);
	
	it('renders correctly', () => {
		expect(fProjectList).toMatchSnapshot();
	});

	it('renders two `FProjectsListItem` items', () => {
		expect(fProjectList.find('FProjectsListItem').length).toEqual(2);
	});

	describe('user filters down `flow` project', () => {
		const filterProjects = shallow(<FProjectsList projects={mockProjectsList} onClickProjectHandler={mockFn} searchValue="flow"/>)

		it('renders only flow project when searchValue equals flow', () => {
			expect(filterProjects.find('FProjectsListItem').dive().find('h4').text()).toEqual('flow');
		});

		it('renders only one project', () => {
			expect(filterProjects.find('FProjectsListItem').length).toEqual(1);	
		});
	});

	
});
