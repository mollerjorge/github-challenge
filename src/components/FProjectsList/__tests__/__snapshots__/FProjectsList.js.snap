// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FProjectsList renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FProjectsList
    onClickProjectHandler={[Function]}
    projects={
        Array [
            Object {
              "id": 1,
              "name": "create-react-app",
              "watchers_count": 1000,
            },
            Object {
              "id": 2,
              "name": "flow",
              "watchers_count": 500,
            },
          ]
    }
    searchValue=""
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <FProjectsListItem
          onClickProjectHandler={[Function]}
          project={
                    Object {
                              "id": 1,
                              "name": "create-react-app",
                              "watchers_count": 1000,
                            }
          }
/>,
        <FProjectsListItem
          onClickProjectHandler={[Function]}
          project={
                    Object {
                              "id": 2,
                              "name": "flow",
                              "watchers_count": 500,
                            }
          }
/>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "onClickProjectHandler": [Function],
          "project": Object {
            "id": 1,
            "name": "create-react-app",
            "watchers_count": 1000,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "onClickProjectHandler": [Function],
          "project": Object {
            "id": 2,
            "name": "flow",
            "watchers_count": 500,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": Symbol(react.fragment),
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <FProjectsListItem
            onClickProjectHandler={[Function]}
            project={
                        Object {
                                    "id": 1,
                                    "name": "create-react-app",
                                    "watchers_count": 1000,
                                  }
            }
/>,
          <FProjectsListItem
            onClickProjectHandler={[Function]}
            project={
                        Object {
                                    "id": 2,
                                    "name": "flow",
                                    "watchers_count": 500,
                                  }
            }
/>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "onClickProjectHandler": [Function],
            "project": Object {
              "id": 1,
              "name": "create-react-app",
              "watchers_count": 1000,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "onClickProjectHandler": [Function],
            "project": Object {
              "id": 2,
              "name": "flow",
              "watchers_count": 500,
            },
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": Symbol(react.fragment),
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
